File: external_research/2026-09-06/robby955_mo_413935_61ac2268b6f9/verification/verify_framed_hadamard_lift_30.py
#!/usr/bin/env python3
"""Exact checks for the framed common-Hadamard order-16 lift.

This verifier separates two statements:

* the displayed order-16 sign matrix has Boolean quadratic maximum 30, so
  F(16) <= 30;
* among lifts with the pinned oriented Hadamard cross frame and arbitrary
  order-four signings on the diagonal, the minimum maximum is exactly 30.

The restricted lower bound is a six-state certificate, not an exhaustive
search over the 64^4 choices of diagonal blocks.
"""

from __future__ import annotations

import hashlib
import itertools
from collections import Counter


if not __debug__:
    raise RuntimeError("assertions must be enabled")


Matrix = tuple[tuple[int, ...], ...]
Spins = tuple[int, ...]

VERTEX_PAIRS = tuple(itertools.combinations(range(4), 2))
CLOUD_PAIRS = VERTEX_PAIRS

ORIENTED_H: Matrix = (
    (1, 1, 1, 1),
    (1, 1, -1, -1),
    (1, -1, 1, -1),
    (-1, 1, 1, -1),
)
BASE_SIGNS = (1, 1, 1, -1, 1, -1)

# Edge order: 01, 02, 03, 12, 13, 23.
P_EDGES = (1, 1, 1, -1, 1, -1)
R_EDGES = (-1, -1, 1, 1, 1, -1)
WITNESS_INTERNAL_EDGES = (P_EDGES, R_EDGES, P_EDGES, R_EDGES)

# Each entry is a (+28 state, -28 state).  The other three clouds have the
# same projective spins.  On cloud 1, the three pairs give the directed cycle
# 0 -> 3 -> 2 -> 0 in PROJECTIVE_REPS.
OBSTRUCTION_PAIRS: tuple[tuple[tuple[Spins, ...], tuple[Spins, ...]], ...] = (
    (
        (
            (1, 1, 1, 1),
            (1, 1, 1, 1),
            (-1, -1, -1, -1),
            (1, 1, 1, -1),
        ),
        (
            (1, 1, 1, 1),
            (-1, -1, 1, 1),
            (-1, -1, -1, -1),
            (-1, -1, -1, 1),
        ),
    ),
    (
        (
            (1, 1, -1, -1),
            (1, 1, -1, -1),
            (-1, -1, 1, 1),
            (1, 1, -1, 1),
        ),
        (
            (1, 1, -1, -1),
            (-1, -1, 1, -1),
            (-1, -1, 1, 1),
            (-1, -1, 1, -1),
        ),
    ),
    (
        (
            (1, 1, -1, -1),
            (1, 1, -1, 1),
            (-1, -1, 1, 1),
            (1, 1, -1, 1),
        ),
        (
            (1, 1, -1, -1),
            (-1, -1, -1, -1),
            (-1, -1, 1, 1),
            (-1, -1, 1, -1),
        ),
    ),
)

EXPECTED_FULL_HISTOGRAM = (
    (-30, 38),
    (-28, 88),
    (-26, 186),
    (-24, 456),
    (-22, 670),
    (-20, 1056),
    (-18, 1634),
    (-16, 1904),
    (-14, 2518),
    (-12, 2960),
    (-10, 3050),
    (-8, 3640),
    (-6, 3774),
    (-4, 4088),
    (-2, 4514),
    (0, 4384),
    (2, 4514),
    (4, 4088),
    (6, 3774),
    (8, 3640),
    (10, 3050),
    (12, 2960),
    (14, 2518),
    (16, 1904),
    (18, 1634),
    (20, 1056),
    (22, 670),
    (24, 456),
    (26, 186),
    (28, 88),
    (30, 38),
)
EXPECTED_HISTOGRAM_SHA256 = (
    "875e4f931630501ec7730abad70df7ab029602f7a48b4b10d56df9de1e319388"
)
EXPECTED_MATRIX_SHA256 = (
    "352392c57458568ddbf2920d4cb487f67d21fb491ab8c75c6862e9c7fc6a9181"
)

PROJECTIVE_REPS = tuple(
    (1,) + tail for tail in itertools.product((1, -1), repeat=3)
)


def edge_matrix(edges: tuple[int, ...]) -> Matrix:
    matrix = [[0] * 4 for _ in range(4)]
    for value, (left, right) in zip(edges, VERTEX_PAIRS):
        matrix[left][right] = value
        matrix[right][left] = value
    return tuple(tuple(row) for row in matrix)


def transpose(matrix: Matrix) -> Matrix:
    return tuple(
        tuple(matrix[row][column] for row in range(len(matrix)))
        for column in range(len(matrix[0]))
    )


def matmul(left: Matrix, right: Matrix) -> Matrix:
    return tuple(
        tuple(
            sum(
                left[row][inner] * right[inner][column]
                for inner in range(len(right))
            )
            for column in range(len(right[0]))
        )
        for row in range(len(left))
    )


def is_hadamard(matrix: Matrix) -> bool:
    product = matmul(matrix, transpose(matrix))
    return product == tuple(
        tuple(4 if row == column else 0 for column in range(4))
        for row in range(4)
    )


def projective_spins(order: int):
    for mask in range(1 << (order - 1)):
        yield (1,) + tuple(
            -1 if (mask >> (coordinate - 1)) & 1 else 1
            for coordinate in range(1, order)
        )


def projective_key(spins: Spins) -> Spins:
    sign = spins[0]
    return tuple(sign * value for value in spins)


def quadratic_energy(matrix: Matrix, spins: Spins) -> int:
    return sum(
        matrix[left][right] * spins[left] * spins[right]
        for left in range(len(matrix))
        for right in range(left + 1, len(matrix))
    )


def internal_energy(edges: tuple[int, ...], spins: Spins) -> int:
    return sum(
        value * spins[left] * spins[right]
        for value, (left, right) in zip(edges, VERTEX_PAIRS)
    )


def cross_energy(clouds: tuple[Spins, ...]) -> int:
    return sum(
        BASE_SIGNS[pair_index]
        * sum(
            clouds[left_cloud][row]
            * ORIENTED_H[row][column]
            * clouds[right_cloud][column]
            for row in range(4)
            for column in range(4)
        )
        for pair_index, (left_cloud, right_cloud) in enumerate(CLOUD_PAIRS)
    )


def block_energy(clouds: tuple[Spins, ...]) -> int:
    return cross_energy(clouds) + sum(
        internal_energy(edges, clouds[cloud])
        for cloud, edges in enumerate(WITNESS_INTERNAL_EDGES)
    )


def assemble_matrix(internal_edges: tuple[tuple[int, ...], ...]) -> Matrix:
    matrix = [[0] * 16 for _ in range(16)]
    for cloud, edges in enumerate(internal_edges):
        block = edge_matrix(edges)
        for row in range(4):
            for column in range(4):
                matrix[4 * cloud + row][4 * cloud + column] = block[row][column]
    for pair_index, (left_cloud, right_cloud) in enumerate(CLOUD_PAIRS):
        sign = BASE_SIGNS[pair_index]
        for row in range(4):
            for column in range(4):
                value = sign * ORIENTED_H[row][column]
                matrix[4 * left_cloud + row][4 * right_cloud + column] = value
                matrix[4 * right_cloud + column][4 * left_cloud + row] = value
    return tuple(tuple(row) for row in matrix)


def verify_structure(matrix: Matrix) -> None:
    if not is_hadamard(ORIENTED_H):
        raise AssertionError("the pinned cross frame is not Hadamard")
    if ORIENTED_H == transpose(ORIENTED_H):
        raise AssertionError("the pinned cross frame unexpectedly became symmetric")
    if any(len(row) != 16 for row in matrix):
        raise AssertionError("assembled matrix is not 16 by 16")
    for row in range(16):
        for column in range(16):
            if matrix[row][column] != matrix[column][row]:
                raise AssertionError("assembled matrix is not symmetric")
            expected_values = (0,) if row == column else (-1, 1)
            if matrix[row][column] not in expected_values:
                raise AssertionError("assembled matrix is not an admissible signing")


def verify_order_four_blocks() -> None:
    optimum = min(
        max(
            abs(internal_energy(edges, spins))
            for spins in projective_spins(4)
        )
        for edges in itertools.product((1, -1), repeat=6)
    )
    if optimum != 4:
        raise AssertionError(("recomputed F(4)", optimum))
    for label, edges in (("P", P_EDGES), ("R", R_EDGES)):
        maximum = max(
            abs(internal_energy(edges, spins))
            for spins in projective_spins(4)
        )
        if maximum != optimum:
            raise AssertionError((label, maximum, optimum))

    # R_ij = s_i s_j P_{perm(i),perm(j)}.
    permutation = (1, 0, 3, 2)
    switching = (1, -1, -1, -1)
    p_matrix = edge_matrix(P_EDGES)
    r_matrix = edge_matrix(R_EDGES)
    transformed = tuple(
        tuple(
            switching[row]
            * switching[column]
            * p_matrix[permutation[row]][permutation[column]]
            for column in range(4)
        )
        for row in range(4)
    )
    if transformed != r_matrix:
        raise AssertionError("the P-to-R switching-permutation certificate failed")


def verify_six_state_obstruction() -> None:
    directed_edges: list[tuple[int, int]] = []
    for positive, negative in OBSTRUCTION_PAIRS:
        if cross_energy(positive) != 28 or cross_energy(negative) != -28:
            raise AssertionError("an obstruction state has the wrong cross energy")
        positive_keys = tuple(projective_key(cloud) for cloud in positive)
        negative_keys = tuple(projective_key(cloud) for cloud in negative)
        if any(
            positive_keys[cloud] != negative_keys[cloud]
            for cloud in (0, 2, 3)
        ):
            raise AssertionError("the paired states do not cancel other blocks")
        directed_edges.append(
            (
                PROJECTIVE_REPS.index(positive_keys[1]),
                PROJECTIVE_REPS.index(negative_keys[1]),
            )
        )
    if tuple(directed_edges) != ((0, 3), (3, 2), (2, 0)):
        raise AssertionError(("obstruction cycle", directed_edges))

    # If a full lift had maximum at most 28, each pair would imply
    # q_D(source) <= q_D(target).  The directed cycle would make the three
    # response values equal.  But q_D(rep_0)-q_D(rep_2) is twice a sum of
    # three signs, hence cannot be zero.
    for edges in itertools.product((1, -1), repeat=6):
        q0 = internal_energy(edges, PROJECTIVE_REPS[0])
        q2 = internal_energy(edges, PROJECTIVE_REPS[2])
        q3 = internal_energy(edges, PROJECTIVE_REPS[3])
        odd_sum_formula = 2 * (edges[1] + edges[3] + edges[5])
        if q0 - q2 != odd_sum_formula:
            raise AssertionError("the odd-sum response identity failed")
        if odd_sum_formula == 0:
            raise AssertionError("a sum of three signs vanished")
        if q0 == q3 == q2:
            raise AssertionError("the impossible response equality occurred")


def common_internal_profile() -> Counter[int]:
    states: list[tuple[tuple[int, ...], int]] = []
    for spins in projective_spins(16):
        clouds = tuple(spins[4 * cloud : 4 * cloud + 4] for cloud in range(4))
        response_indices = tuple(
            PROJECTIVE_REPS.index(projective_key(cloud)) for cloud in clouds
        )
        states.append((response_indices, cross_energy(clouds)))

    profile: Counter[int] = Counter()
    for edges in itertools.product((1, -1), repeat=6):
        responses = tuple(
            internal_energy(edges, representative)
            for representative in PROJECTIVE_REPS
        )
        maximum = max(
            abs(cross + sum(responses[index] for index in indices))
            for indices, cross in states
        )
        profile[maximum] += 1
    return profile


def verify_witness(matrix: Matrix) -> tuple[str, int]:
    matrix_digest = hashlib.sha256(repr(matrix).encode()).hexdigest()
    if matrix_digest != EXPECTED_MATRIX_SHA256:
        raise AssertionError(("matrix digest", matrix_digest))
    projective_histogram: Counter[int] = Counter()
    cross_maximum = 0
    for spins in projective_spins(16):
        clouds = tuple(spins[4 * cloud : 4 * cloud + 4] for cloud in range(4))
        direct = quadratic_energy(matrix, spins)
        blocked = block_energy(clouds)
        if direct != blocked:
            raise AssertionError("direct and blockwise energy computations disagree")
        projective_histogram[direct] += 1
        cross_maximum = max(cross_maximum, abs(cross_energy(clouds)))

    full_histogram = tuple(
        (energy, 2 * count) for energy, count in sorted(projective_histogram.items())
    )
    digest = hashlib.sha256(repr(list(full_histogram)).encode()).hexdigest()
    if full_histogram != EXPECTED_FULL_HISTOGRAM:
        raise AssertionError("the complete energy histogram changed")
    if digest != EXPECTED_HISTOGRAM_SHA256:
        raise AssertionError(("energy histogram digest", digest))
    maximum = max(abs(energy) for energy in projective_histogram)
    projective_maximizers = sum(
        count
        for energy, count in projective_histogram.items()
        if abs(energy) == maximum
    )
    if maximum != 30 or projective_maximizers != 38 or cross_maximum != 28:
        raise AssertionError(
            ("witness maxima", maximum, projective_maximizers, cross_maximum)
        )
    return digest, projective_maximizers


def verify_corruption_controls(matrix: Matrix) -> None:
    corrupted_h = [list(row) for row in ORIENTED_H]
    corrupted_h[0][0] *= -1
    if is_hadamard(tuple(tuple(row) for row in corrupted_h)):
        raise AssertionError("Hadamard corruption was not detected")

    corrupted_matrix = [list(row) for row in matrix]
    corrupted_matrix[0][4] *= -1
    if all(
        corrupted_matrix[row][column] == corrupted_matrix[column][row]
        for row in range(16)
        for column in range(16)
    ):
        raise AssertionError("transpose corruption was not detected")

    corrupted_pair = [
        [list(cloud) for cloud in state]
        for state in OBSTRUCTION_PAIRS[0]
    ]
    corrupted_pair[0][0][1] *= -1
    positive = tuple(tuple(cloud) for cloud in corrupted_pair[0])
    negative = tuple(tuple(cloud) for cloud in corrupted_pair[1])
    if cross_energy(positive) == 28 and all(
        projective_key(positive[cloud]) == projective_key(negative[cloud])
        for cloud in (0, 2, 3)
    ):
        raise AssertionError("six-state corruption was not detected")


def main() -> None:
    matrix = assemble_matrix(WITNESS_INTERNAL_EDGES)
    verify_structure(matrix)
    verify_order_four_blocks()
    verify_six_state_obstruction()
    histogram_digest, projective_maximizers = verify_witness(matrix)

    uniform_profile = common_internal_profile()
    expected_uniform_profile = Counter({42: 32, 40: 18, 46: 8, 38: 6})
    if uniform_profile != expected_uniform_profile:
        raise AssertionError(("common-internal profile", uniform_profile))
    verify_corruption_controls(matrix)

    print("order=16")
    print("cross_frame=common_oriented_H4")
    print("base_signs=+,+,+,-,+,-")
    print("internal_pattern=P,R,P,R")
    print("internal_maxima=4,4,4,4")
    print("P_R_switching_permutation_equivalent=TRUE")
    print("projective_spins_checked=32768")
    print("cross_only_maximum=28")
    print("lift_maximum=30")
    print(f"projective_maximizers={projective_maximizers}")
    print("full_maximizers=76")
    print(f"matrix_sha256={EXPECTED_MATRIX_SHA256}")
    print(f"energy_histogram_sha256={histogram_digest}")
    print("six_state_obstruction_cycle=0->3->2->0")
    print("fixed_H_arbitrary_internal_minimum=30")
    print("common_literal_internal_minimum=38")
    print("common_literal_internal_profile=M38:6,M40:18,M42:32,M46:8")
    print("corruption_controls=hadamard_entry,transpose,six_state_pair")
    print("framed_hadamard_lift_30_verification=PASSED")


if __name__ == "__main__":
    main()

File: external_research/2026-09-06/robby955_mo_413935_61ac2268b6f9/verification/verify_framed_hadamard_lift_30.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

enum { CLOUDS = 4, CLOUD_SIZE = 4, ORDER = 16, STATES = 32768 };

static const int h[4][4] = {
    {1, 1, 1, 1},
    {1, 1, -1, -1},
    {1, -1, 1, -1},
    {-1, 1, 1, -1},
};
static const int pairs[6][2] = {
    {0, 1}, {0, 2}, {0, 3}, {1, 2}, {1, 3}, {2, 3},
};
static const int base[6] = {1, 1, 1, -1, 1, -1};
static const int p_edges[6] = {1, 1, 1, -1, 1, -1};
static const int r_edges[6] = {-1, -1, 1, 1, 1, -1};
static const int *const internal[4] = {
    p_edges, r_edges, p_edges, r_edges,
};

static const int obstruction[6][4][4] = {
    {{1, 1, 1, 1}, {1, 1, 1, 1}, {-1, -1, -1, -1}, {1, 1, 1, -1}},
    {{1, 1, 1, 1}, {-1, -1, 1, 1}, {-1, -1, -1, -1}, {-1, -1, -1, 1}},
    {{1, 1, -1, -1}, {1, 1, -1, -1}, {-1, -1, 1, 1}, {1, 1, -1, 1}},
    {{1, 1, -1, -1}, {-1, -1, 1, -1}, {-1, -1, 1, 1}, {-1, -1, 1, -1}},
    {{1, 1, -1, -1}, {1, 1, -1, 1}, {-1, -1, 1, 1}, {1, 1, -1, 1}},
    {{1, 1, -1, -1}, {-1, -1, -1, -1}, {-1, -1, 1, 1}, {-1, -1, 1, -1}},
};

static int edge_energy(const int edges[6], const int spins[4]) {
    int result = 0;
    for (int edge = 0; edge < 6; ++edge) {
        const int left = pairs[edge][0];
        const int right = pairs[edge][1];
        result += edges[edge] * spins[left] * spins[right];
    }
    return result;
}

static int cross_energy(const int clouds[4][4]) {
    int result = 0;
    for (int edge = 0; edge < 6; ++edge) {
        const int left = pairs[edge][0];
        const int right = pairs[edge][1];
        int block = 0;
        for (int row = 0; row < 4; ++row) {
            for (int column = 0; column < 4; ++column) {
                block += clouds[left][row] * h[row][column]
                         * clouds[right][column];
            }
        }
        result += base[edge] * block;
    }
    return result;
}

static int block_energy(const int spins[16]) {
    int clouds[4][4];
    int result;
    for (int cloud = 0; cloud < 4; ++cloud) {
        for (int coordinate = 0; coordinate < 4; ++coordinate) {
            clouds[cloud][coordinate] = spins[4 * cloud + coordinate];
        }
    }
    result = cross_energy(clouds);
    for (int cloud = 0; cloud < 4; ++cloud) {
        result += edge_energy(internal[cloud], clouds[cloud]);
    }
    return result;
}

static void assemble(int matrix[16][16]) {
    for (int row = 0; row < 16; ++row) {
        for (int column = 0; column < 16; ++column) {
            matrix[row][column] = 0;
        }
    }
    for (int cloud = 0; cloud < 4; ++cloud) {
        for (int edge = 0; edge < 6; ++edge) {
            const int left = 4 * cloud + pairs[edge][0];
            const int right = 4 * cloud + pairs[edge][1];
            matrix[left][right] = internal[cloud][edge];
            matrix[right][left] = internal[cloud][edge];
        }
    }
    for (int edge = 0; edge < 6; ++edge) {
        const int left_cloud = pairs[edge][0];
        const int right_cloud = pairs[edge][1];
        for (int row = 0; row < 4; ++row) {
            for (int column = 0; column < 4; ++column) {
                const int left = 4 * left_cloud + row;
                const int right = 4 * right_cloud + column;
                const int value = base[edge] * h[row][column];
                matrix[left][right] = value;
                matrix[right][left] = value;
            }
        }
    }
}

static int direct_energy(const int matrix[16][16], const int spins[16]) {
    int result = 0;
    for (int left = 0; left < 16; ++left) {
        for (int right = left + 1; right < 16; ++right) {
            result += matrix[left][right] * spins[left] * spins[right];
        }
    }
    return result;
}

static void decode_projective(unsigned int mask, int spins[16]) {
    spins[0] = 1;
    for (int coordinate = 1; coordinate < 16; ++coordinate) {
        const unsigned int bit = 1U << (unsigned int)(coordinate - 1);
        spins[coordinate] = (mask & bit) != 0U ? -1 : 1;
    }
}

static int projective_index(const int spins[4]) {
    const int sign = spins[0];
    int index = 0;
    for (int coordinate = 1; coordinate < 4; ++coordinate) {
        index *= 2;
        if (sign * spins[coordinate] == -1) {
            index += 1;
        }
    }
    return index;
}

static int projectively_equal(const int left[4], const int right[4]) {
    const int sign = left[0] * right[0];
    for (int coordinate = 0; coordinate < 4; ++coordinate) {
        if (left[coordinate] != sign * right[coordinate]) {
            return 0;
        }
    }
    return 1;
}

static int verify_hadamard(void) {
    for (int left = 0; left < 4; ++left) {
        for (int right = 0; right < 4; ++right) {
            int product = 0;
            for (int coordinate = 0; coordinate < 4; ++coordinate) {
                product += h[left][coordinate] * h[right][coordinate];
            }
            if (product != (left == right ? 4 : 0)) {
                return 0;
            }
        }
    }
    return 1;
}

static void representative(int index, int spins[4]) {
    spins[0] = 1;
    for (int coordinate = 3; coordinate >= 1; --coordinate) {
        spins[coordinate] = (index & 1) != 0 ? -1 : 1;
        index >>= 1;
    }
}

static void edges_from_mask(unsigned int mask, int edges[6]) {
    for (int edge = 0; edge < 6; ++edge) {
        const unsigned int bit = 1U << (unsigned int)edge;
        edges[edge] = (mask & bit) != 0U ? -1 : 1;
    }
}

static void verify_obstruction(void) {
    const int expected_sources[3] = {0, 3, 2};
    const int expected_targets[3] = {3, 2, 0};
    for (int pair = 0; pair < 3; ++pair) {
        const int *positive_flat = &obstruction[2 * pair][0][0];
        const int *negative_flat = &obstruction[2 * pair + 1][0][0];
        const int (*positive)[4] = (const int (*)[4])positive_flat;
        const int (*negative)[4] = (const int (*)[4])negative_flat;
        if (cross_energy(positive) != 28 || cross_energy(negative) != -28) {
            fprintf(stderr, "obstruction cross energy mismatch\n");
            exit(EXIT_FAILURE);
        }
        for (int cloud = 0; cloud < 4; ++cloud) {
            if (cloud != 1
                && !projectively_equal(positive[cloud], negative[cloud])) {
                fprintf(stderr, "obstruction cancellation mismatch\n");
                exit(EXIT_FAILURE);
            }
        }
        if (projective_index(positive[1]) != expected_sources[pair]
            || projective_index(negative[1]) != expected_targets[pair]) {
            fprintf(stderr, "obstruction cycle mismatch\n");
            exit(EXIT_FAILURE);
        }
    }

    for (unsigned int mask = 0U; mask < 64U; ++mask) {
        int edges[6];
        int u0[4];
        int u2[4];
        int u3[4];
        int q0;
        int q2;
        int q3;
        edges_from_mask(mask, edges);
        representative(0, u0);
        representative(2, u2);
        representative(3, u3);
        q0 = edge_energy(edges, u0);
        q2 = edge_energy(edges, u2);
        q3 = edge_energy(edges, u3);
        if (q0 - q2 != 2 * (edges[1] + edges[3] + edges[5])) {
            fprintf(stderr, "odd-sum identity mismatch\n");
            exit(EXIT_FAILURE);
        }
        if (q0 == q3 && q3 == q2) {
            fprintf(stderr, "impossible response equality occurred\n");
            exit(EXIT_FAILURE);
        }
    }
}

static void verify_order_four(void) {
    int optimum = 100;
    int p_maximum = 0;
    int r_maximum = 0;
    for (unsigned int mask = 0U; mask < 64U; ++mask) {
        int edges[6];
        int maximum = 0;
        edges_from_mask(mask, edges);
        for (int index = 0; index < 8; ++index) {
            int spins[4];
            int energy;
            representative(index, spins);
            energy = abs(edge_energy(edges, spins));
            if (energy > maximum) {
                maximum = energy;
            }
        }
        if (maximum < optimum) {
            optimum = maximum;
        }
    }
    for (int index = 0; index < 8; ++index) {
        int spins[4];
        int p_energy;
        int r_energy;
        representative(index, spins);
        p_energy = abs(edge_energy(p_edges, spins));
        r_energy = abs(edge_energy(r_edges, spins));
        if (p_energy > p_maximum) {
            p_maximum = p_energy;
        }
        if (r_energy > r_maximum) {
            r_maximum = r_energy;
        }
    }
    if (optimum != 4 || p_maximum != 4 || r_maximum != 4) {
        fprintf(stderr, "order-four maximum mismatch\n");
        exit(EXIT_FAILURE);
    }
}

static void verify_common_internal_profile(void) {
    static int cross_values[STATES];
    static uint8_t response_indices[STATES][4];
    int profile[47] = {0};
    const int expected[47] = {
        [38] = 6, [40] = 18, [42] = 32, [46] = 8,
    };
    for (unsigned int mask = 0U; mask < (1U << 15U); ++mask) {
        int spins[16];
        int clouds[4][4];
        decode_projective(mask, spins);
        for (int cloud = 0; cloud < 4; ++cloud) {
            for (int coordinate = 0; coordinate < 4; ++coordinate) {
                clouds[cloud][coordinate] = spins[4 * cloud + coordinate];
            }
            response_indices[mask][cloud]
                = (uint8_t)projective_index(clouds[cloud]);
        }
        cross_values[mask] = cross_energy(clouds);
    }
    for (unsigned int edge_mask = 0U; edge_mask < 64U; ++edge_mask) {
        int edges[6];
        int responses[8];
        int maximum = 0;
        edges_from_mask(edge_mask, edges);
        for (int index = 0; index < 8; ++index) {
            int spins[4];
            representative(index, spins);
            responses[index] = edge_energy(edges, spins);
        }
        for (unsigned int state = 0U; state < (1U << 15U); ++state) {
            int energy = cross_values[state];
            for (int cloud = 0; cloud < 4; ++cloud) {
                energy += responses[response_indices[state][cloud]];
            }
            energy = abs(energy);
            if (energy > maximum) {
                maximum = energy;
            }
        }
        if (maximum >= 47) {
            fprintf(stderr, "common-internal maximum outside profile\n");
            exit(EXIT_FAILURE);
        }
        profile[maximum] += 1;
    }
    for (int maximum = 0; maximum < 47; ++maximum) {
        if (profile[maximum] != expected[maximum]) {
            fprintf(stderr, "common-internal profile mismatch\n");
            exit(EXIT_FAILURE);
        }
    }
}

int main(void) {
    int matrix[16][16];
    int histogram[241] = {0};
    int maximum = 0;
    int maximizers = 0;
    int cross_maximum = 0;
    const int expected_counts[31] = {
        38, 88, 186, 456, 670, 1056, 1634, 1904, 2518, 2960, 3050,
        3640, 3774, 4088, 4514, 4384, 4514, 4088, 3774, 3640, 3050,
        2960, 2518, 1904, 1634, 1056, 670, 456, 186, 88, 38,
    };

    if (!verify_hadamard()) {
        fprintf(stderr, "Hadamard check failed\n");
        return EXIT_FAILURE;
    }
    assemble(matrix);
    for (int row = 0; row < 16; ++row) {
        for (int column = 0; column < 16; ++column) {
            if (matrix[row][column] != matrix[column][row]
                || (row == column && matrix[row][column] != 0)
                || (row != column && abs(matrix[row][column]) != 1)) {
                fprintf(stderr, "admissibility check failed\n");
                return EXIT_FAILURE;
            }
        }
    }

    for (unsigned int mask = 0U; mask < (1U << 15U); ++mask) {
        int spins[16];
        int clouds[4][4];
        int direct;
        int blocked;
        int cross;
        decode_projective(mask, spins);
        direct = direct_energy(matrix, spins);
        blocked = block_energy(spins);
        if (direct != blocked) {
            fprintf(stderr, "direct/block energy mismatch\n");
            return EXIT_FAILURE;
        }
        histogram[direct + 120] += 1;
        if (abs(direct) > maximum) {
            maximum = abs(direct);
        }
        for (int cloud = 0; cloud < 4; ++cloud) {
            for (int coordinate = 0; coordinate < 4; ++coordinate) {
                clouds[cloud][coordinate] = spins[4 * cloud + coordinate];
            }
        }
        cross = abs(cross_energy(clouds));
        if (cross > cross_maximum) {
            cross_maximum = cross;
        }
    }
    for (int energy = -30; energy <= 30; energy += 2) {
        const int expected = expected_counts[(energy + 30) / 2];
        if (2 * histogram[energy + 120] != expected) {
            fprintf(stderr, "full histogram mismatch at energy %d\n", energy);
            return EXIT_FAILURE;
        }
        if (abs(energy) == maximum) {
            maximizers += histogram[energy + 120];
        }
    }
    if (maximum != 30 || maximizers != 38 || cross_maximum != 28) {
        fprintf(stderr, "maximum summary mismatch\n");
        return EXIT_FAILURE;
    }

    verify_order_four();
    verify_obstruction();
    verify_common_internal_profile();

    puts("projective_spins_checked=32768");
    puts("direct_block_histogram_match=TRUE");
    puts("lift_maximum=30");
    puts("projective_maximizers=38");
    puts("cross_only_maximum=28");
    puts("six_state_obstruction=PASSED");
    puts("common_literal_internal_minimum=38");
    puts("strict_c_framed_hadamard_lift_30_verification=PASSED");
    return EXIT_SUCCESS;
}
